Make statement patterns anchored

mjambon opened the pull request:
This is for https://github.com/returntocorp/semgrep/issues/5472

So far, I added test cases for:
* JavaScript
* Bash
* Dockerfile

The following change makes the tests above pass:
```
   | G.ExprStmt (a1, a2), B.ExprStmt (b1, b2) ->
-      m_expr_deep a1 b1 >>= fun () -> m_tok a2 b2
+      m_expr a1 b1 >>= fun () -> m_tok a2 b2
```
but a bunch of other tests fail. I haven't looked into why they're failing and I'm not sure how this should be resolved.

PR checklist:

- [ ] Documentation is up-to-date
- [ ] Changelog is up-to-date
- [ ] Change has no security implications (otherwise, ping security team)

Commit:
Prevent a statement pattern 'a;' from matching the target 'f(a);'

Commit:
Add test cases for anchored statement matching for Bash
(for https://github.com/returntocorp/semgrep/issues/5472)

Commit:
Add a test for anchored shell command matching
(for https://github.com/returntocorp/semgrep/issues/5472)

Commit:
Improve test cases

mjambon commented:
I'm handing this off to @aryx because I'm going on vacation.

aryx committed:
Merge branch 'develop' into mj-bash-and

aryx committed:
Merge branch 'develop' into mj-bash-and

aryx commented:
"a bunch of other test fails", actually 122 tests fail, so it introduces too many regressions.
Including some of the tests you wrote before @mjambon like 
```
FROM debian
...
RUN apt-get ...
```
that you want to match
```
# ERROR:
FROM debian:testing
RUN echo hello
RUN apt-get update && apt-get install -y fortune
```

aryx commented:
I'll make an option so one can disable this behavior on a per-rule basis.

aryx committed:
Add new options: implicit_deep_exprstmt, add more tests

Also cleanup how we handl implicit deep exprstmt

test plan:
test file included
make test

aryx committed:
changelog
